### DISCUSSION TITLE
Arithmetic function `isinfinite` missing?

### DISCUSSION THREAD

**Discussion by dnewhall** [ksh93/ksh#798] [Q&A]
The man page says `isinfinite` is one of the predefined arithmetic functions, but when I go and test it:
```shell
$ echo $(( isinfinite(inf) ))
ksh: isinfinite(inf) : unknown function
```

`isinf(inf)` works as expected, though.

Was this removed and the documentation needs to be updated? Should it be in there and no one noticed it was missing until now?

**Comment by McDutchie** [comment]
As the manual page says: “Any of the following math library functions _that are in the C math library_ can be used within an arithmetic expression:”, etc.

What functions are in the C math library depends on the operating system. Your system has `isinf` but not `isinfinite` (as does mine; I don't know on what systems `isinfinite` exists). Both are checked for at compile time [in math.tab](https://github.com/ksh93/ksh/blob/caf8a66f3ecbc549b5550bcb7565050bce221d1a/src/cmd/ksh93/data/math.tab#L57-L58), which is parsed by [math.sh](https://github.com/ksh93/ksh/blob/caf8a66f3ecbc549b5550bcb7565050bce221d1a/src/cmd/ksh93/features/math.sh#L1), which is [executed by iffe](https://github.com/ksh93/ksh/blob/caf8a66f3ecbc549b5550bcb7565050bce221d1a/src/cmd/ksh93/Mamfile#L41-L49).

**Comment by dnewhall** [comment]
Should Ksh make some guarantees here? They are really useful, but if they are platform-defined it makes any shell script that uses them inherently non-portable.

I just found out this is a big issue with `fpclassify`, which cannot be portably used at all in its current state (`fpclassify(nan)` returns 0 on Linux, 1 on MacOS, and 2 on FreeBSD...)

**Comment by McDutchie** [comment]
It would be nice if ksh came with its own fallback implementations, but I don't think anyone has the time or inclination to make that happen.

With `fpclassify`(3) you're of course supposed to use the classification macros to check the result (`FP_INFINITE`, `FP_NAN`, etc.), otherwise it's not portable.

**Comment by dnewhall** [comment]
The more I think about this, the more I think Ksh really should guarantee these functions are provided (or at least some)

I was curious which functions might not be provided on different platforms based on their standard conformance, since (in theory) Ksh can be built with some elbow grease on any platform that has a C90 compiler. Here's what I was able to find. All of these exist on Linux, but I don't know about BSD and MacOS (I can check later tonight)

C90:
Trig: sin, cos, tan
Inverse trig: asin, acos, atan, atan2
Hyperbolic trig: sinh, cosh, tanh
Exponentiation: exp, ldexp
Logorithms: log, log10
Power: pow, sqrt
Rounding: ceil, fabs, floor, fmod
Integer: abs

C99:
Basic: fdim, fmin, fmax, remainder, trunc, rint
Inverse hyperbolic trig: asinh, acosh, atanh
Exponentiation: exp2, exp10, expm1
Logarithms: log2, log10, log1p
Power: cbrt, hypot
Rounding: nearbyint, round
Error functions: erf, erfc
Float manipulation: ilogb, logb, copysign, nextafter, nexttoward
Classification: fpclassify, isfinite, isinf, isnan, isnormal, isunordered, isless, isgreater, islessqual, isgreaterequal, signbit

TS 18661-1:2014:
iszero, issubnormal

TS 18661-4:2015:
exp10

BSD functions:
isinf
finite

SVID:
j0, j1, jn, y0, y1, yn

Unknown:
isinfinite

**Reply by dnewhall** [reply]
Just did a build on MacOS: It has all the functions that Linux has, except the ones from the Technical Specifications: iszero, issubnormal, and exp10. Also doesn't have isinfinite, as well.

Also, a small note on the non-portability of `fpclassify`:
- On Linux: 0 = NaN, 1 = infinity, 2 = zero, 3 = subnormal, 4 = normal.
- On Solaris and (at least) FreeBSD: 1 = Infinite, 2 = NaN, 4 = normal, 8 = subnormal, 16 = zero.
- On MacOS: 1 = NaN, 2 = infinity, 3 = zero, 4 = normal, 5 = subnormal, 6 = supernormal (PowerPC only).

**Comment by dnewhall** [comment]
@McDutchie, Would you be amenable to me killing math.tab and having math.sh checking for functions support as a group?

This is what I would change and document:
- All C90 functions (plus `int` and `float`) are always guaranteed to be provided (as required from the C lib standards from 34 years ago)
- All C99 functions will be provided if a C99 compiler is used.
- The currently used BSD functions, the TS functions, and `isinfinite` will be guaranteed to be provided if the C99 functions are available (with emulated versions generated if unavailable).
- The Bessel functions will be tested for independently and provided if available.
- `fpclassify` will be officially deprecated since it's inherently non-portable (it will still be available as a function, but not in the man page anymore for a few releases)
- I then want to add a few more TS 18661-1:2014 / C23 mathematical functions *only if we can emulate them using the C99 ones* (specifically `roundeven`, `sinpi` .. `atanpi`, `expNm1`, `logNp1`)

**Reply by McDutchie** [reply]
Yes, I would be open to that, and I would really appreciate your contribution. It sounds like a good project for the next version, 93u+m/1.1 (the current dev branch). For the 1.0 branch, it's too late for major changes like that.

The dev branch will require C99 at some point in the not too distant future (see #777), so I would not bother with C90 compatibility.

As for `fpclassify`, since it's perfectly portable in C as long as you use the FP_* classification macros, I suggest we hardcode a numeric return value for every classification macro to make it the same across platforms. We should hardcode the values that are currently returned on Linux, as that is the most popular platform for ksh use.

Note that #771 is really the top priority issue when it comes to maths. I don't know how any fix/redesign for that issue would affect your contribution because I haven't figured out how to fix that.
